### PR TITLE
[Internal] Change Feed: Adds query support for change feed

### PR DESCRIPTION
**Pull request by ajparule**
# Pull Request Template

## Description

This PR adds query support on change feed operation.
The query will be part of request body and will be a POST operation on dbs/{db1}/colls/{{col1}/docs endpoint.

New methods are added in ContainerInternal for this feature and are internal.
public abstract FeedIterator GetChangeFeedStreamIteratorWithQuery
public abstract FeedIterator<T> GetChangeFeedIteratorWithQuery<T>

ChangeFeedQuerySpec will contain the sql query and should be passed to above methods.

Example :
Insert 3  documents as below
{ "id" : "id1", "city":"city1", "pkey" : "pkey1" }
{ "id" : "id2", "city":"city2", "pkey" : "pkey1" }
{ "id" : "id3", "city":"city3", "pkey" : "pkey1" }

Start CF with below query :
"select * from c where c.city = 'city3'"

Change feed response will return below records which match the query
{ "id" : "id3", "city":"city3", "pkey" : "pkey1" }

Testing:
Added test for change Feed query on both current and previous image using direct and gateway modes.

## Type of change
- [] New feature (non-breaking change which adds functionality)

## Closing issues

To automatically close an issue: closes #IssueNumber

### CODE REVIEW

File: Microsoft.Azure.Cosmos/src/ChangeFeed/ChangeFeedQuerySpec.cs [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+        {
+            this.QueryText = queryText;
+        }
+
+        /// <summary>
+        /// Gets or sets the text of the Azure Cosmos DB database query.
+        /// </summary>
+        /// <value>The text of the database query.</value>
+        [DataMember(Name = "query")]
+        public string QueryText { get; set; }
```
Comment (ealsur) [inline Microsoft.Azure.Cosmos/src/ChangeFeed/ChangeFeedQuerySpec.cs:51]: why is the public setter on this properties needed?

File: Microsoft.Azure.Cosmos/src/ChangeFeed/ChangeFeedQuerySpec.cs [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+
+        /// <summary>
+        /// Gets or sets if database query should be run on previous image if present.
+        /// </summary>
+        /// <value>True if query should be run on previous image. False, otherwise.</value>
+        [DataMember(Name = "enableQueryOnPreviousImage")]
+        internal bool EnableQueryOnPreviousImage { get; set; }
```
Comment (ealsur) [inline Microsoft.Azure.Cosmos/src/ChangeFeed/ChangeFeedQuerySpec.cs:58]: The class is already internal, no need to set this property also as internal.

File: Microsoft.Azure.Cosmos/src/ChangeFeed/Pagination/CrossPartitionChangeFeedAsyncEnumerator.cs [hunk 3]
```diff
@@ -9,6 +9,7 @@ namespace Microsoft.Azure.Cosmos.ChangeFeed.Pagination
     using System.Threading;
     using System.Threading.Tasks;
     using Microsoft.Azure.Cosmos.Pagination;
+    using Microsoft.Azure.Cosmos.Query.Core;
```
Comment (ealsur) [inline Microsoft.Azure.Cosmos/src/ChangeFeed/Pagination/CrossPartitionChangeFeedAsyncEnumerator.cs:12]: nit: Is this change needed?

File: Microsoft.Azure.Cosmos/src/ChangeFeedHelper.cs [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+// Copyright (c) Microsoft Corporation.  All rights reserved.
+//------------------------------------------------------------
+
+namespace Microsoft.Azure.Cosmos
+{
+    using Microsoft.Azure.Documents;
+
+    internal static class ChangeFeedHelper
+    {
+        internal static bool isChangeFeedWithQueryRequest(
```
Comment (ealsur) [inline Microsoft.Azure.Cosmos/src/ChangeFeedHelper.cs:11]: Method is not private, so the casing can be upper camel case

```suggestion
        internal static bool IsChangeFeedWithQueryRequest(
```

File: Microsoft.Azure.Cosmos/src/ChangeFeedHelper.cs [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+           OperationType operationType,
+           bool hasStreamPayload = false)
+        {
+            if (operationType == OperationType.ReadFeed && hasStreamPayload)
```
Comment (ealsur) [inline Microsoft.Azure.Cosmos/src/ChangeFeedHelper.cs:15]: This could be simplified by just:

```suggestion
            return operationType == OperationType.ReadFeed && hasStreamPayload;
```

File: Microsoft.Azure.Cosmos/src/ChangeFeedHelper.cs [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+
+namespace Microsoft.Azure.Cosmos
+{
+    using Microsoft.Azure.Documents;
+
+    internal static class ChangeFeedHelper
+    {
+        internal static bool isChangeFeedWithQueryRequest(
+           OperationType operationType,
+           bool hasStreamPayload = false)
```
Comment (ealsur) [inline Microsoft.Azure.Cosmos/src/ChangeFeedHelper.cs:13]: This is a new method, it would be better not to use defaults

File: Microsoft.Azure.Cosmos/src/Pagination/NetworkAttachedDocumentContainer.cs [hunk 7]
```diff
@@ -286,6 +288,13 @@ internal sealed class NetworkAttachedDocumentContainer : IMonadicDocumentContain
                 throw new ArgumentNullException(nameof(changeFeedPaginationOptions));
             }
 
+            MemoryStream memoryStream = null;
+            if (changeFeedPaginationOptions.changeFeedQuerySpec != null && changeFeedPaginationOptions.changeFeedQuerySpec.QueryText != null)
+            {
+                string queryText = Newtonsoft.Json.JsonConvert.SerializeObject(changeFeedPaginationOptions.changeFeedQuerySpec);
+                memoryStream = new MemoryStream(Encoding.UTF8.GetBytes(queryText));
```
Comment (ealsur) [inline Microsoft.Azure.Cosmos/src/Pagination/NetworkAttachedDocumentContainer.cs:295]: We should be doing serialization through the CosmosSerializer interface, similar to what happens for the SqlQuerySpec: https://github.com/Azure/azure-cosmos-dotnet-v3/blob/8093426f90a9f31cb4a185fb0ffc134353e6738b/Microsoft.Azure.Cosmos/src/Serializer/CosmosSerializerCore.cs#L84

```
internal Stream ToStreamChangeFeedQuerySpec(ChangeFeedQuerySpec input)
{
	return CosmosSerializerCore.propertiesSerializer.ToStream(input);
}
```

Comment (ealsur) [reply Microsoft.Azure.Cosmos/src/Pagination/NetworkAttachedDocumentContainer.cs:295]: Reference to the CosmosQueryClient: https://github.com/Azure/azure-cosmos-dotnet-v3/blob/8093426f90a9f31cb4a185fb0ffc134353e6738b/Microsoft.Azure.Cosmos/src/Query/v3Query/CosmosQueryClientCore.cs#L94

File: Microsoft.Azure.Cosmos/src/ChangeFeed/ChangeFeedQuerySpec.cs [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+//------------------------------------------------------------
+
+namespace Microsoft.Azure.Cosmos.ChangeFeed
+{
+    using System;
+    using System.Runtime.Serialization;
+    using Microsoft.Azure.Cosmos.Query.Core;
+
+    [DataContract]
+    internal sealed class ChangeFeedQuerySpec
```
Comment (ealsur) [inline Microsoft.Azure.Cosmos/src/ChangeFeed/ChangeFeedQuerySpec.cs:12]: Should this be public?

File: Microsoft.Azure.Cosmos/src/ChangeFeed/ChangeFeedQuerySpec.cs [hunk 9]
```diff
@@ -0,0 +1,68 @@
...
+        /// <summary>
+        /// Initializes a new instance of the <see cref="T:Microsoft.Azure.Documents.ChangeFeedQuerySpec"/> class for the Azure Cosmos DB service.</summary>
+        /// <param name="queryText">The text of the database query.</param>
+        /// <param name="enableQueryOnPreviousImage">The boolean when enabaled runs database query on previous image.</param>
+        /// <remarks> 
+        /// The default constructor initializes any fields to their default values.
+        /// </remarks>
+        public ChangeFeedQuerySpec(string queryText, bool enableQueryOnPreviousImage)
+        {
+            this.QueryText = queryText;
```
Comment (ealsur) [inline Microsoft.Azure.Cosmos/src/ChangeFeed/ChangeFeedQuerySpec.cs:33]: Null validations

File: Microsoft.Azure.Cosmos/src/ChangeFeed/ChangeFeedQuerySpec.cs [hunk 10]
```diff
@@ -0,0 +1,68 @@
...
+        /// </remarks>
+        public ChangeFeedQuerySpec()
+            : this(null)
+        {
+        }
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="T:Microsoft.Azure.Documents.ChangeFeedQuerySpec"/> class for the Azure Cosmos DB service.</summary>
+        /// <param name="queryText">The text of the database query.</param>
+        /// <param name="enableQueryOnPreviousImage">The boolean when enabaled runs database query on previous image.</param>
```
Comment (ealsur) [inline Microsoft.Azure.Cosmos/src/ChangeFeed/ChangeFeedQuerySpec.cs:27]: What is query on previous image? What are the implications of setting it? Does it only work on certain conditions like Full Fidelity or also on Inremental?

File: Microsoft.Azure.Cosmos/src/ChangeFeed/ChangeFeedQuerySpec.cs [hunk 11]
```diff
@@ -0,0 +1,68 @@
...
+        /// Gets or sets if database query should be run on previous image if present.
+        /// </summary>
+        /// <value>True if query should be run on previous image. False, otherwise.</value>
+        [DataMember(Name = "enableQueryOnPreviousImage")]
+        internal bool EnableQueryOnPreviousImage { get; set; }
+
+        /// <summary>
+        /// Returns a value that indicates whether <see cref="EnableQueryOnPreviousImage"/> property should be serialized.
+        /// </summary>
+        internal bool ShouldSerializeEnableQueryOnPreviousImage()
```
Comment (ealsur) [inline Microsoft.Azure.Cosmos/src/ChangeFeed/ChangeFeedQuerySpec.cs:63]: What is the usage of this method?

File: Microsoft.Azure.Cosmos/src/ChangeFeed/ChangeFeedQuerySpec.cs (hunk 10 above)
Comment (ajparule) [reply Microsoft.Azure.Cosmos/src/ChangeFeed/ChangeFeedQuerySpec.cs:27]: enableQueryOnPreviousImage when set will run the query on the previous image in BE and return the change feed document if previous image satisfies the query.
It works only with Full Fidelity and where previous image is persisted for replace and delete operations, currently this is enabled for materializedviews source collections

File: Microsoft.Azure.Cosmos/src/Resource/Container/ContainerCore.cs [hunk 12]
```diff
@@ -626,5 +627,73 @@ public override async Task<CollectionRoutingMap> GetRoutingMapAsync(Cancellation
               trace: trace,
               cancellationToken: cancellationToken);
         }
+
+        public override FeedIterator GetChangeFeedStreamIteratorWithQuery(
+           ChangeFeedStartFrom changeFeedStartFrom,
+           ChangeFeedMode changeFeedMode,
+           ChangeFeedRequestOptions changeFeedRequestOptions = null,
+           ChangeFeedQuerySpec changeFeedQuerySpec = null)
```
[the remark this answers is not in the document]
Comment (ajparule) [reply Microsoft.Azure.Cosmos/src/Resource/Container/ContainerCore.cs:635]: "Why not add the query to the ChangeFeedRequestOptions? What is the point of adding a new public API for which the only important parameter is optional?"
--We want to keep this feature internal and will be only used by MaterializedViews builder application code. Hence added a new internal API.

 What happens if a customer creates a Change Feed iterator with one query, reads a couple of pages, fetches the continuation, then at some point later creates a new iterator with the continuation (https://docs.microsoft.com/en-us/azure/cosmos-db/sql/change-feed-pull-model?tabs=dotnet#saving-continuation-tokens) and passes a different query?
 --In this case the change feed iterator will use the new query and will start from the LSN specified in the continuation token

**Comment by ajparule** [comment]
> Is this API for internal consumption? Is it on preview?
> This API is for internal consumption only and will be used by materializdViews builder. 

> Do you have an API design discussion or proposal? 
No.
> Which were the alternatives to "GetChangeFeedStreamIteratorWithQuery" that were evaluated and why was "GetChangeFeedStreamIteratorWithQuery" selected?
 Since Currently there is no plan to expose this for public use , went with a separate internal API.
 Currently there is no plan to public expose the CF with query.

> Can you share usage examples on the Description?
Will update the description

File: Microsoft.Azure.Cosmos/src/ChangeFeedHelper.cs [hunk 13]
```diff
@@ -0,0 +1,25 @@
...
+        {
+            return operationType == OperationType.ReadFeed && hasStreamPayload;
+            {
+                // ChangeFeed with payload is a CF with query support and will
+                // be a query POST request.
+                return true;
+            }
+            return false;
```
Comment (abhijitpai) [inline Microsoft.Azure.Cosmos/src/ChangeFeedHelper.cs:21]: Fix

File: Microsoft.Azure.Cosmos/src/GatewayStoreClient.cs [hunk 14]
```diff
@@ -255,6 +256,12 @@ private static async Task<Stream> BufferContentIfAvailableAsync(HttpResponseMess
             {
                 httpMethod = HttpMethod.Post;
             }
+            else if (ChangeFeedHelper.isChangeFeedWithQueryRequest(request.OperationType, request.Body != null))
```
Comment (abhijitpai) [inline Microsoft.Azure.Cosmos/src/GatewayStoreClient.cs:259]: >i [](http://example.com/codeflow?start=38&length=1)

Fix

File: Microsoft.Azure.Cosmos/src/ChangeFeed/Pagination/ChangeFeedPaginationOptions.cs [hunk 15]
```diff
@@ -30,14 +31,18 @@ internal sealed class ChangeFeedPaginationOptions : PaginationOptions
...
+            ChangeFeedQuerySpec changeFeedQuerySpec = null)
             : base(pageSizeHint, jsonSerializationFormat, additionalHeaders)
         {
             this.Mode = mode ?? throw new ArgumentNullException(nameof(mode));
+            this.changeFeedQuerySpec = changeFeedQuerySpec ?? new ChangeFeedQuerySpec();
         }
 
         public ChangeFeedMode Mode { get; }
 
+        public ChangeFeedQuerySpec changeFeedQuerySpec { get; }
```
Comment (abhijitpai) [inline Microsoft.Azure.Cosmos/src/ChangeFeed/Pagination/ChangeFeedPaginationOptions.cs:44]: >c [](http://example.com/codeflow?start=35&length=1)

C

File: Microsoft.Azure.Cosmos/src/Resource/Container/ContainerCore.cs (hunk 12 above)
[the remark this answers is not in the document]
Comment (abhijitpai) [reply Microsoft.Azure.Cosmos/src/Resource/Container/ContainerCore.cs:635]: Can we make the changeFeedQuerySpec as mandatory on these calls then?

File: Microsoft.Azure.Cosmos/src/Resource/Container/ContainerCore.cs [hunk 16]
```diff
@@ -626,5 +627,73 @@ public override async Task<CollectionRoutingMap> GetRoutingMapAsync(Cancellation
...
+
+            if (changeFeedMode == null)
+            {
+                throw new ArgumentNullException(nameof(changeFeedMode));
+            }
+
+            NetworkAttachedDocumentContainer networkAttachedDocumentContainer = new NetworkAttachedDocumentContainer(
+                this,
+                this.queryClient,
+                Guid.NewGuid(),
```
Comment (abhijitpai) [inline Microsoft.Azure.Cosmos/src/Resource/Container/ContainerCore.cs:650]: >  Gui [](http://example.com/codeflow?start=14&length=5)

Add paramName: for clarity

File: Microsoft.Azure.Cosmos/src/ChangeFeedHelper.cs [hunk 17]
```diff
@@ -0,0 +1,25 @@
...
+
+namespace Microsoft.Azure.Cosmos
+{
+    using Microsoft.Azure.Documents;
+
+    internal static class ChangeFeedHelper
+    {
+        internal static bool IsChangeFeedWithQueryRequest(
+           OperationType operationType,
+           bool hasStreamPayload = false)
```
Comment (abhijitpai) [inline Microsoft.Azure.Cosmos/src/ChangeFeedHelper.cs:13]: >m [](http://example.com/codeflow?start=24&length=1)

hasPayload is sufficient

File: Microsoft.Azure.Cosmos/src/ChangeFeed/ChangeFeedQuerySpec.cs (hunk 11 above)
Comment (abhijitpai) [reply Microsoft.Azure.Cosmos/src/ChangeFeed/ChangeFeedQuerySpec.cs:63]: Remove if unused

File: Microsoft.Azure.Cosmos/src/ChangeFeed/ChangeFeedQuerySpec.cs (hunk 8 above)
Comment (abhijitpai) [reply Microsoft.Azure.Cosmos/src/ChangeFeed/ChangeFeedQuerySpec.cs:12]: We want to start with internal so we don't change the public surface.

File: Microsoft.Azure.Cosmos/src/ChangeFeed/ChangeFeedQuerySpec.cs (hunk 10 above)
Comment (ealsur) [reply Microsoft.Azure.Cosmos/src/ChangeFeed/ChangeFeedQuerySpec.cs:27]: In that case then, if the user starts a non-FF change feed, and they set this flag, will the request fail with an explanatory error?

File: Microsoft.Azure.Cosmos/src/Resource/Container/ContainerCore.cs (hunk 12 above)
[the remark this answers is not in the document]
Comment (ealsur) [reply Microsoft.Azure.Cosmos/src/Resource/Container/ContainerCore.cs:635]: > We want to keep this feature internal and will be only used by MaterializedViews builder application code. Hence added a new internal API.

You can make this an internal property in ChangeFeedRequestOptions and achieve the same thing.

If you are adding a new Method to the Container class, then it needs to be easy for users to understand from its signature, which is the usecase they need to use it with.

Will this work with any Change Feed Mode (From Beginning, Some specific Time, Now?) and with any FeedRange (A Partition Key value, or a partition?)

Even if this is initially an internal API, it will eventually graduate to be public, so we need to think the API design that it will eventually have.

File: Microsoft.Azure.Cosmos/src/Handler/RequestInvokerHandler.cs [hunk 18]
```diff
@@ -280,6 +281,12 @@ internal class RequestInvokerHandler : RequestHandler
                         request.Headers.ContentType = RuntimeConstants.MediaTypes.JsonPatch;
                     }
 
+                    if (ChangeFeedHelper.isChangeFeedWithQueryRequest(operationType, streamPayload != null))
```
Comment (ealsur) [inline Microsoft.Azure.Cosmos/src/Handler/RequestInvokerHandler.cs:284]: Could you describe the flow for a scenario where the query does not match with any existing documents? Right now change feed has 2 responses: HTTP 304 means no new changes at the moment (the user can choose to sleep or stop and continue later), or HTTP 200 means new changes.

What happens if the query does not have any matching documents? How does the user identify this scenario and what is the Continuation (Etag) produced by the backend?

File: Microsoft.Azure.Cosmos/src/ChangeFeed/ChangeFeedQuerySpec.cs (hunk 10 above)
Comment (ajparule) [reply Microsoft.Azure.Cosmos/src/ChangeFeed/ChangeFeedQuerySpec.cs:27]: Yes below exception is thrown

Response status code does not indicate success: BadRequest (400); Substatus: 0; ActivityId: 0d68861e-350c-43a6-9f80-5d1f2419f167; Reason: (Response status code does not indicate success: BadRequest (400); Substatus: 0; ActivityId: 0d68861e-350c-43a6-9f80-5d1f2419f167; Reason: (Response status code does not indicate success: BadRequest (400); Substatus: 0; ActivityId: 0d68861e-350c-43a6-9f80-5d1f2419f167; Reason: (Message: {"Errors":["Property 'enableQueryFilterOnPreviousImage' can only be used with Full-Fidelity Change Feed."]}

File: Microsoft.Azure.Cosmos/src/Handler/RequestInvokerHandler.cs (hunk 18 above)
Comment (ajparule) [reply Microsoft.Azure.Cosmos/src/Handler/RequestInvokerHandler.cs]: If the query does not match any document and there are more documents to be searched then user receives 200 with feedResponse count set to 0 and the continuation token will have the last searched LSN

{"type":"continuation","value":"\"7126\""}
.....
.....
{"type":"continuation","value":"\"13997\""}

Once all the documents are searched 304 code is returned with last searched LSN

Comment (ealsur) [reply Microsoft.Azure.Cosmos/src/Handler/RequestInvokerHandler.cs]: Thanks for the clarification. 
This means then that users can receive empty pages on Change Feed, which should be properly documented as it has not been a case previously. Our documentation shows users how to consume the Change Feed (https://docs.microsoft.com/en-us/azure/cosmos-db/sql/change-feed-pull-model?tabs=dotnet#consuming-an-entire-containers-changes) and users would be expecting to receive items, but they might not when using a Query.

**Review by ealsur — COMMENTED** [review]
Unblocking the PR after discussion, please address the pending comments (like the serialization)

File: Microsoft.Azure.Cosmos/src/Resource/Container/ContainerCore.cs (hunk 12 above)
[the remark this answers is not in the document]
Comment (ajparule) [reply Microsoft.Azure.Cosmos/src/Resource/Container/ContainerCore.cs:635]: Since the CF query is part of request body its not a a good idea to add it in ChangeFeedRequestOptions .

Will this work with any Change Feed Mode (From Beginning, Some specific Time, Now?) and with any FeedRange (A Partition Key value, or a partition?)
-->Yes

Even if this is initially an internal API, it will eventually graduate to be public, so we need to think the API design that it will eventually have.
--> public API will have a different wire format which is yet to be designed.

File: Microsoft.Azure.Cosmos/src/ChangeFeed/ChangeFeedQuerySpec.cs (hunk 9 above)
Comment (ealsur) [reply Microsoft.Azure.Cosmos/src/ChangeFeed/ChangeFeedQuerySpec.cs:33]: The Spec can be an optional parameter sure, but the Text on the Spec is not, right?

File: Microsoft.Azure.Cosmos/src/Resource/Container/ContainerCore.cs (hunk 12 above)
[the remark this answers is not in the document]
Comment (ealsur) [reply Microsoft.Azure.Cosmos/src/Resource/Container/ContainerCore.cs:635]: Echoing what @abhijitpai said then, can we make the parameter mandatory? Otherwise there is no distinction between this method and the currently existing one, both can be used to consume the Change Feed without a query if I wanted

File: Microsoft.Azure.Cosmos/src/Handler/RequestInvokerHandler.cs [hunk 19]
```diff
@@ -298,7 +305,8 @@ internal class RequestInvokerHandler : RequestHandler
 
         internal static HttpMethod GetHttpMethod(
             ResourceType resourceType,
-            OperationType operationType)
+            OperationType operationType,
+            bool hasStreamPayload = false)
```
Comment (abhijitpai) [inline Microsoft.Azure.Cosmos/src/Handler/RequestInvokerHandler.cs:309]: >tream [](http://example.com/codeflow?start=21&length=5)

hasPayload

**Review by abhijitpai — APPROVED** [review]
:shipit:

**Comment by ajparule** [comment]
> Is this API for internal consumption? Is it on preview?
> Its for Internal consumption only

> Do you have an API design discussion or proposal? Which were the alternatives to "GetChangeFeedStreamIteratorWithQuery" that were evaluated and why was "GetChangeFeedStreamIteratorWithQuery" selected?
> 

> Can you share usage examples on the Description?
updated the description

**Review by ealsur — APPROVED** [review]
